operators/pkg/tenant-controller/tenant_controller.go

```diff
@@ -144,27 +148,23 @@ func (r *TenantReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		retrigErr = err
 	}
 
+	// Determine if the personal namespace should be deleted
+
+	nsName := fmt.Sprintf("tenant-%s", strings.ReplaceAll(tn.Name, ".", "-"))
+
+	// Test if namespace has been open for too long; attempt to delete if there are no instances inside
+	keepNsOpen, err := r.enforceNamespaceKeepAliveOrDelete(ctx, &tn, nsName, r.TenantWorkspaceKeepAlive)
+	if err != nil {
+		klog.Errorf("Error in r.List: unable to capture instances in tenant workspace %s -> %s", nsName, err)
+	}
+
 	// update resource quota in the status of the tenant after checking validity of workspaces.
 	tn.Status.Quota = forge.TenantResourceList(workspaces, tn.Spec.Quota)
 
-	nsName := fmt.Sprintf("tenant-%s", strings.ReplaceAll(tn.Name, ".", "-"))
-	nsOk, err := r.createOrUpdateClusterResources(ctx, &tn, nsName)
-	if nsOk {
-		klog.Infof("Namespace %s for tenant %s updated", nsName, tn.Name)
-		tn.Status.PersonalNamespace.Created = true
-		tn.Status.PersonalNamespace.Name = nsName
-		if err != nil {
-			klog.Errorf("Unable to update cluster resource of tenant %s -> %s", tn.Name, err)
-			retrigErr = err
-			tnOpinternalErrors.WithLabelValues("tenant", "cluster-resources").Inc()
-		}
-		klog.Infof("Cluster resourcess for tenant %s updated", tn.Name)
-	} else {
-		klog.Errorf("Unable to update namespace of tenant %s -> %s", tn.Name, err)
-		tn.Status.PersonalNamespace.Created = false
-		tn.Status.PersonalNamespace.Name = ""
-		retrigErr = err
-		tnOpinternalErrors.WithLabelValues("tenant", "cluster-resources").Inc()
+	var nsOk bool
+	nsOk, err = r.enforceClusterResources(ctx, &tn, nsName, keepNsOpen)
+	if err != nil {
+		klog.Errorf("Error when enforcing cluster resources for tenant %s -> %s", tn.Name, err)
```

Comment:
Same here about returning in case of error.

Reply:
Fixed in [c8b2bef](https://github.com/netgroup-polito/CrownLabs/pull/821/commits/c8b2bef303c5f7a8322434f6106ea6635319408a)